TuyaMCU Precise Temperature

jaymemaurice opened the pull request:
Sensor packetValue now interpreted as float based on configuration of  TempRes command.

## Description:

**Related issue (if applicable):** fixes #10917

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] The code change is tested and works on Tasmota core ESP8266 V.2.7.4.9
  - [ ] The code change is tested and works on Tasmota core ESP32 V.1.0.5-rc6
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

jaymemaurice committed:
TuyaMCU Precise Temperature

Sensor packetValue now interpreted as float based on configuration of  TempRes command.
Beware TempRes default may be 1 thus a TuyaMCU device may need TempRes set to 0 after this change if it's currently working.
See https://github.com/arendst/Tasmota/issues/10917

jaymemaurice commented:
Can confirm this may break hass which appears to have built in assumptions for TuyaMCU devices
![image](https://user-images.githubusercontent.com/38895319/107681852-9c08f580-6c6d-11eb-9369-d00bf476a3ca.png)

Jason2866 commented:
Dont use pow. Tasmota has its own pow function. Take a look into support_float.ino

ascillato commented:
> Can confirm this may break hass which appears to have built in assumptions for TuyaMCU devices

Sorry, what do you see that is wrong?

jaymemaurice commented:
@ascillato - actually not HASS. Seems somewhere I am missing the division. Sometime some MQTT messages are still coming as integer.
![image](https://user-images.githubusercontent.com/38895319/107683001-0ff7cd80-6c6f-11eb-8e31-6d882a10423b.png)
I was looking at wrong MQTT topic. I will debug and look into support_float as directed

jaymemaurice commented:
@Jason2866 in this case I am using 4 possibilities of small numbers: 10^3, 10^2, 10^1, 10^0...
I looked into the FastPow and PreciseFastPow algorithims and wonder if either are really necessary here. 
If looking for optimization, a function with switch/case might be better... would that be preferred?

ascillato commented:
The preference is what uses less flash and ram. Using pow() increases a lot both flash and ram usage.

jaymemaurice committed:
Precise temperature without pow

Implement TuyaAdjustedTemperature which takes precision to offset the integer value. 
Using TempRes parameter as these should align

jaymemaurice committed:
Fix missing divisor

Fix missing divisior

jaymemaurice commented:
Standby, I have one more commit

jaymemaurice committed:
Fix loss of precision 

Fix loss of precision regression

jaymemaurice committed:
Update xdrv_16_tuyamcu.ino

Change type uint8_t to uint16_t on TuyaAdjustedTemperature to support temperatures over 25.5 or 255 degrees.

jaymemaurice commented:
Barring any new suggestions from anyone, I am done with this and I am content in the manner the device is working.
